framework/resource/internal/wrapper_test.go

```diff
@@ -0,0 +1,212 @@
+			t.Fatalf("test %d: unexpected error = %#v", i, err)
+		}
+
+		if underlying.Name() != "testNoopResource" {
+			t.Fatalf("test %d: expected %q resource, got %q", i, "testNoopResource", underlying.Name())
+		}
+	}
+}
+
+// Test_Underlying_NonWrapped tests Underlying returns the same resource when
+// it isn't a Wrapper.
+func Test_Underlying_NonWrapped(t *testing.T) {
+	r := &testNoopResource{}
+
+	underlying, err := Underlying(r)
+	if err != nil {
+		t.Fatalf("unexpected error = %#v", err)
+	}
+
+	if underlying != r {
+		t.Fatalf("expected %#v, got %#v", r, underlying)
+	}
+}
+
+// Test_Underlying_Loop tests if Underlying panics when there is a loop.
+func Test_Underlying_Loop(t *testing.T) {
+	testCases := []struct {
+		ResourceFunc func() framework.Resource
+	}{
+		// Test 0. r1 -> r1.
+		{
+			ResourceFunc: func() framework.Resource {
+				r1 := &testWrappingResource{}
+
+				r1.resource = r1
+
+				return r1
+			},
+		},
+		// Test 1. r1 -> r2 -> r3 -> r1.
+		{
+			ResourceFunc: func() framework.Resource {
+				r1 := &testWrappingResource{}
+				r2 := &testWrappingResource{}
+				r3 := &testWrappingResource{}
+
+				r1.resource = r2
+				r2.resource = r3
+				r3.resource = r1
+
+				return r1
+			},
+		},
+		// Test 1. r1 -> r2 -> r3 -> r4 -> r2.
+		{
+			ResourceFunc: func() framework.Resource {
+				r1 := &testWrappingResource{}
+				r2 := &testWrappingResource{}
+				r3 := &testWrappingResource{}
+				r4 := &testWrappingResource{}
+
+				r1.resource = r2
+				r2.resource = r3
+				r3.resource = r4
+				r4.resource = r2
+
+				return r1
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		_, err := Underlying(tc.ResourceFunc())
+
+		if !IsLoopDetected(err) {
+			t.Fatalf("test %d: expected %v, got %v", i, loopDetectedError, err)
+		}
+	}
+}
+
+type testNoopResource struct{}
+
+func (r *testNoopResource) Name() string {
+	return "testNoopResource"
+}
+
+func (r *testNoopResource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	return nil, nil
+}
+
+func (r *testNoopResource) GetDesiredState(ctx context.Context, obj interface{}) (interface{}, error) {
+	return nil, nil
+}
+
+func (r *testNoopResource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
+	return nil, nil
+}
+
+func (r *testNoopResource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
+	return nil, nil
+}
+
+func (r *testNoopResource) ApplyCreateChange(ctx context.Context, obj, createChange interface{}) error {
+	return nil
+}
+
+func (r *testNoopResource) ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error {
+	return nil
+}
+
+func (r *testNoopResource) ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error {
+	return nil
+}
+
+type testWrappingResource struct {
+	resource framework.Resource
+}
+
+func (r *testWrappingResource) Wrapped() framework.Resource {
+	return r.resource
+}
```

Comment:
The only difference between the resources is this one has `Wrapped()` methods which fulfills `Wrapper` interface.